python/openinference-instrumentation/src/openinference/instrumentation/__init__.py

```diff
@@ -12,11 +33,228 @@ def suppress_tracing() -> Iterator[None]:
             # No tracing will occur within this block
             ...
     """
-    try:
-        from opentelemetry.context import _SUPPRESS_INSTRUMENTATION_KEY, attach, detach, set_value
-    except ImportError:
-        yield
+
+    def __enter__(self) -> "suppress_tracing":
+        self._token = attach(set_value(_SUPPRESS_INSTRUMENTATION_KEY, True))
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        """Raise any exception triggered within the runtime context."""
+        detach(self._token)
+        return None
+
+
+class UsingAttributes:
+    slots = [
+        "_token",
+        "_session_id",
+        "_user_id",
+        "_metadata",
+        "_tags",
+    ]
+
+    def __init__(
+        self,
+        session_id: str = "",
+        user_id: str = "",
+        metadata: Dict[str, str] = {},
+        tags: List[str] = [],
+    ) -> None:
+        if not isinstance(session_id, str):
+            raise TypeError("session_id must be a string")
+        if not isinstance(user_id, str):
+            raise TypeError("user_id must be a string")
+        if not is_dict_of(metadata, key_allowed_types=(str), value_allowed_types=(str)):
+            raise TypeError("metadata must be a dictionary with string keys and values")
+        if not is_list_of(tags, str):
+            raise TypeError("tags must be a list of strings")
+
+        self._session_id = session_id
+        self._user_id = user_id
+        self._metadata = metadata
+        self._tags = tags
+
+    def attach_context(self) -> None:
+        ctx = get_current()
+        if self._session_id:
+            ctx = set_value(SpanAttributes.SESSION_ID, self._session_id, ctx)
+        if self._user_id:
+            ctx = set_value(SpanAttributes.USER_ID, self._user_id, ctx)
+        if self._metadata:
+            ctx = set_value(SpanAttributes.METADATA, self._metadata, ctx)
+        if self._tags:
+            ctx = set_value(SpanAttributes.TAG_TAGS, self._tags, ctx)
+
+        self._token = attach(ctx)
+        return
+
+    def detach_context(self) -> None:
+        detach(self._token)
+        return
+
+
+class using_session(AbstractContextManager, UsingAttributes):
```

Comment:
I am not sure we need this level of abstraction using abstract base classes and multiple inheritance to start? A class or function is probably enough. What is this level of abstraction buying us in this case?

Reply:
Removed the abstract classes inheritance